controller/tests/update_test_helpers.ml

```diff
@@ -0,0 +1,366 @@
+    let expected_outcome = slot_spec_to_outcome case in
+    let expected_outcome_str =
+            (sexp_of_expected_outcomes expected_outcome |> Sexplib.Sexp.to_string)
+    in
+    let test_case_descr =
+        Format.sprintf
+            "%s\t->\t%s"
+            (slot_spec_to_string case)
+            expected_outcome_str
+    in
+    Alcotest_lwt.test_case test_case_descr `Quick (fun _ () ->
+        let mocks = init_test_deps () in
+
+        let () = setup_mocks_from_system_slot_spec mocks case in
+
+        let module UpdateServiceI = (val mocks.update_service) in
+        let%lwt out_state = UpdateServiceI.run_step GettingVersionInfo in
+        if state_matches_expected_outcome out_state expected_outcome then
+            Lwt.return ()
+        else
+            Alcotest.fail (Format.sprintf
+                "Reached state [%s] does not match expected outcome [%s]"
+                (statefmt out_state)
+                expected_outcome_str
+            )
+    )
+
+
+(* === Update scenario spec framework === *)
+
+type action_descr = string
+type action_check = Update.state -> bool Lwt.t
+type mock_update = unit -> unit
+
+type scenario_spec =
+  | StateReached of Update.state
+  | ActionDone of action_descr * action_check
+  | UpdateMock of mock_update
+
+let specfmt spec = match spec with
+    | StateReached s -> "StateReached: " ^ (statefmt s);
+    | ActionDone (descr, c) -> "ActionDone: " ^ descr;
+    | UpdateMock _ -> "UpdateMock: <fun>"
+
+let _MAGIC_PAT = "<..>"
+
+(* string equality, but the magic patern `<..>` is treated
+  as a placeholder for any sub-string. The implementation converts the
+  `expected` string to a regex where the magic pattern is replaced with ".*",
+  while being careful to `Str.quote` the rest of the string to not accidentally
+  treat them as regex expressions.
+*)
+let str_match_with_magic_pat expected actual =
+    let open Str in
+    let magic_pattern = regexp_string _MAGIC_PAT in
+    let exp_parts = full_split magic_pattern expected in
+    let exp_regexp = regexp @@ String.concat "" @@ List.map (fun (p) ->
+        match p with
+            | Text a -> quote a
+            | Delim _ -> ".*"
+    ) exp_parts in
+    string_match exp_regexp actual 0
+
+
+let state_formatter out inp = Format.fprintf out "%s" (statefmt inp)
+
+let t_state =
+    let state_to_str s =
+        Update.sexp_of_state s
+            (* using _hum instead of _mach, because _mach seems to remove
+               whitespace between atoms in some cases *)
+            |> Sexplib.Sexp.to_string_hum
+            (* remove new lines to ignore whitespace differences *)
+            |> Str.global_replace (Str.regexp_string "\n") ""
+    in
+    let state_eq expected actual =
+        (expected == actual) || (
+            str_match_with_magic_pat
+                (* Using string repr is a horrible hack, but avoids having to
+                   pattern match on every variant in the state ADT *)
```

Comment:
> Why would we have to match on every variant in the ADT, could you provide an example?

Because we need to define equality for `state`, which would require defining equality for all the cases, including e.g. `UpToDate of version_info`.

Reply:
> Also, I don’t understand at this point where the magic pattern is created, is it in Update.sexp_of_state?

Do you mean it's unclear how it's used? Here are a few places:
- https://github.com/dividat/playos/blob/18318766cc308beb9a84e6493986b46f84281db5/controller/tests/update_tests.ml#L65
- https://github.com/dividat/playos/blob/18318766cc308beb9a84e6493986b46f84281db5/controller/tests/update_tests.ml#L77
- https://github.com/dividat/playos/blob/18318766cc308beb9a84e6493986b46f84281db5/controller/tests/update_tests.ml#L12-L14

Reply:
Could be renamed to `_MAGIC_WILDCARD_PAT` or just `_WILDCARD_PAT` without the magic :)